src/Core/Services/MetadataProviders/MetadataProviderFactory.cs

```diff
@@ -56,13 +56,24 @@ private void ConfigureMetadataProviders()
                     _ => throw new NotSupportedException(dataSource.DatabaseTypeNotSupportedMessage),
                 };
 
-                _metadataProviders.Add(dataSourceName, metadataProvider);
+                // In a hot reload scenario, the metadata might be updated dynamically. The below code will ensure that:
+                // 1. The _metadataProviders dictionary does not end up with duplicate keys, which would cause an exception.
+                // 2. The metadata provider for a given data source is always up-to-date, reflecting the latest configuration.
+                if (_metadataProviders.ContainsKey(dataSourceName))
+                {
+                    _metadataProviders[dataSourceName] = metadataProvider;
+                }
+                else
+                {
+                    _metadataProviders.Add(dataSourceName, metadataProvider);
+                }
             }
         }
 
         public void OnConfigChanged(object? sender, HotReloadEventArgs args)
         {
             ConfigureMetadataProviders();
+            InitializeAsync().Wait();
```

Comment:
There's a lot of work that happens in the `Initialize()` functions of a MetadataProvider such as fetching db object metadata (this may be different if backend DB changes or connection string points to different DB) , then further validation 
https://github.com/Azure/data-api-builder/blob/d9fd18c407f58cc7d7f6e8e0f6b7fa4b0660b7ff/src/Core/Services/MetadataProviders/SqlMetadataProvider.cs#L269-L311